Kernel/FileSystem/FUSE/FUSEConnection.h

```diff
@@ -0,0 +1,134 @@
+
+        // FIXME: Fill in proper values for these.
+        header->uid = 0;
+        header->gid = 0;
+        header->pid = 1;
+
+        u8* payload = bit_cast<u8*>(request->data() + sizeof(fuse_in_header));
+        memcpy(payload, request_body.data(), request_body.size());
+
+        return request;
+    }
+
+    ErrorOr<NonnullOwnPtr<KBuffer>> send_request_and_wait_for_a_reply(int opcode, int nodeid, ReadonlyBytes request_body)
+    {
+        auto* device = bit_cast<FUSEDevice*>(m_description->device());
+
+        // FIXME: Send the init request from the filesystem itself right after it has been
+        // mounted. (Without blocking the mount syscall.)
+        if (!m_initialized)
+            TRY(handle_init());
+
+        auto request = TRY(create_request(opcode, nodeid, m_unique, request_body));
+        auto response = TRY(device->send_request_and_wait_for_a_reply(m_description, request->bytes()));
+
+        if (validate_response(response.ptr(), m_unique++).is_error())
+            return Error::from_errno(EIO);
+
+        return response;
+    }
+
+    ~FUSEConnection()
+    {
+        auto* device = bit_cast<FUSEDevice*>(m_description->device());
+        // Unblock the userspace daemon and tell it to shut down.
+        device->shutdown_for_description(m_description);
+    }
+
+private:
+    FUSEConnection(NonnullRefPtr<OpenFileDescription> description)
+        : m_description(description)
+    {
+    }
+
+    ErrorOr<void> validate_response(KBuffer* response, u32 unique)
+    {
+        fuse_out_header* header = bit_cast<fuse_out_header*>(response->data());
+        if (header->unique != unique) {
+            dmesgln("FUSE: Received a mismatched request");
+            return Error::from_errno(EINVAL);
+        }
+
+        if (header->len > response->size()) {
+            dmesgln("FUSE: Received an excessively large request");
+            return Error::from_errno(EINVAL);
+        }
+
+        return {};
+    }
+
+    ErrorOr<void> handle_init()
+    {
+        fuse_init_in init_request;
+        init_request.major = FUSE_KERNEL_VERSION;
+        init_request.minor = FUSE_KERNEL_MINOR_VERSION;
+        init_request.max_readahead = 512;
+        init_request.flags = 0;
+
+        auto* device = bit_cast<FUSEDevice*>(m_description->device());
+
+        auto request = TRY(create_request(FUSE_INIT, 0, 0, { &init_request, sizeof(init_request) }));
+        auto response = TRY(device->send_request_and_wait_for_a_reply(m_description, request->bytes()));
+
+        if (validate_response(response.ptr(), 0).is_error())
+            return Error::from_errno(EIO);
+
+        fuse_init_out* init = bit_cast<fuse_init_out*>(response->data() + sizeof(fuse_out_header));
+
+        m_major = init->major;
+        m_minor = init->minor;
+
+        m_initialized = true;
+
+        return {};
+    }
+
+    NonnullRefPtr<OpenFileDescription> m_description;
+    bool m_initialized = false;
+    u32 m_unique = 0;
```

Comment:
The usual syntax we choose for class members is something like `u32 m_my_nice_variable { 0 };` :)